cardinal/clustering.py

```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
         kmeans_args['n_clusters'] = batch_size
         super().__init__(KMeans(**kmeans_args), batch_size)
 
+class GMMLikelihoodSampler(BaseQuerySampler):
+    """ GM (Gaussian Mixture Models) based query sampler.
+    In order to increase diversity, it is possible to use maximum likelihood to select samples.
+
+    Args:
+        clustering: A clustering algorithm matching the sklearn interface
+        batch_size: Number of samples to draw when predicting.
+
+    Attributes:
+        clustering_ : The fitted clustering estimator.
+    """
+    def __init__(self, clustering, batch_size):
+        super().__init__(batch_size)
+        self.clustering_ = clustering
+
+    def fit(self, X, y=None) -> 'GMMLikelihoodSampler':
+        """Does nothing, this method is unsupervised.
+        
+        Args:
+            X: Labeled samples of shape (n_samples, n_features).
+            y: Labels of shape (n_samples).
+        
+        Returns:
+            The object itself
+        """
+        return self
+
+    def select_samples(self, X: np.array) -> np.array:
+        """Clusters the samples and select the ones with highest likelihood.
+        
+        Args:
+            X: Pool of unlabeled samples of shape (n_samples, n_features).
+            sample_weight: Weight of the samples of shape (n_samples),
+                optional.
+
+        Returns:
+            Indices of the selected samples of shape (batch_size).
+        """
+        from scipy.stats import multivariate_normal
+        if self._not_enough_samples(X):
+            return np.arange(X.shape[0])
+
+        gmm_model = self.clustering_.fit(X)
+        centers= np.empty(shape=(gmm_model.n_components, 1), dtype=int)
+        print(gmm_model.n_components)
+        for i in range(gmm_model.n_components):
+            try:
+                density = multivariate_normal.logpdf(X, cov=gmm_model.covariances_[i]
+                                                , mean=gmm_model.means_[i]
+                                                , allow_singular=True)
+            except linalg.LinAlgError:
+                cov = gmm_model.covariances_[i]
+                approximated_cov = np.linalg.pinv(cov)
+                density = multivariate_normal.logpdf(X, cov=approximated_cov
+                                                , mean=gmm_model.means_[i]
+                                                , allow_singular=True)
+            centers[i, :] = np.argmax(density) 
+        return centers.reshape(gmm_model.n_components,)
+
+class GMMSampler(GMMLikelihoodSampler):
+    """Select samples as highest likelihood to GMM clusters.
+
+    Args:
+        batch_size: Number of samples to draw when predicting.
+    """
+    def __init__(self, batch_size, **gmm_args):
+        check_modules('sklearn', 'clustering.KmeansSampler')
```

Comment:
This method just checks the version and raises a meaningful error. Not critical, but it's better to have the right class name.

```suggestion
        check_modules('sklearn', 'clustering.GMMSampler')
```